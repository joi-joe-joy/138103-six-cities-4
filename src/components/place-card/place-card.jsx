import React, {PureComponent} from "react";
import IconBookmark from "../../Icons/icon-bookmark.svg";
import pt from "prop-types";
import {HouseType} from "../../const.js";

class PlaceCard extends PureComponent {
  constructor(props) {
    super(props);
    this.handleCardHover = this.handleCardHover.bind(this);
  }

  handleCardHover(card) {
    this.props.onCardHover(card);
  }

  render() {
    const {offer, onBookmarkClick} = this.props;

    return (
      <article className="cities__place-card place-card"
        onMouseOver={this.handleCardHover.bind(this, offer)}>
        {offer.premium &&
          <div className="place-card__mark">
            <span>Premium</span>
          </div>
        }
        <div className="cities__image-wrapper place-card__image-wrapper">
          <a href="#">
            <img className="place-card__image" src={offer.picture} width="260" height="200" alt={offer.title}></img>
          </a>
        </div>
        <div className="place-card__info">
          <div className="place-card__price-wrapper">
            <div className="place-card__price">
              <b className="place-card__price-value">&euro;{offer.price}</b>
              <span className="place-card__price-text">&#47;&nbsp;night</span>
            </div>
            <button
              onClick={onBookmarkClick}
              className="place-card__bookmark-button button"
              type="button">
              <IconBookmark width="17" height="18" />
              <span className="visually-hidden">To bookmarks</span>
            </button>
          </div>
          <div className="place-card__rating rating">
            <div className="place-card__stars rating__stars">
              <span style={{width: `${offer.rating * 20}%`}}></span>
              <span className="visually-hidden">Rating</span>
            </div>
          </div>
          <h2 className="place-card__name">
            <a href="#">{offer.title}</a>
          </h2>
          <p className="place-card__type">{offer.type}</p>
        </div>
      </article>
    );
  }
}

PlaceCard.propTypes = {
  offer: pt.shape({
    title: pt.string.isRequired,
    premium: pt.bool.isRequired,
    picture: pt.string.isRequired,
    price: pt.number.isRequired,
    rating: pt.number.isRequired,
    type: pt.oneOf([HouseType.APARTMENT, HouseType.ROOM, HouseType.HOUSE, HouseType.HOTEL])
  }).isRequired,
  onBookmarkClick: pt.func.isRequired,
  onCardHover: pt.func.isRequired
};

export default PlaceCard;
