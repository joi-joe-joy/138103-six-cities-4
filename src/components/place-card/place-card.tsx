import * as React from "react";
import classnames from "classnames";
import {Link} from "react-router-dom";
import {HouseTypeTemplate, AppRoute} from "../../const";
import {PlaceCardType, Offer} from "../../types";
import withFavorite from "../../hocs/with-favorite/with-favorite";
import ButtonFavorite from "../button-favorite/button-favorite";

interface Props {
  offer: Offer;
  onCardHover?: (offer: Offer) => void;
  onCardHoverOut?: () => void;
  type: PlaceCardType.CITIES | PlaceCardType.FAVORITES | PlaceCardType.NEAR | PlaceCardType.PROPERTY;
}

const ButtonFavoriteWrap = withFavorite(ButtonFavorite);

class PlaceCard extends React.PureComponent<Props, {}> {
  constructor(props) {
    super(props);
    this.handleCardHover = this.handleCardHover.bind(this);
    this.handleCardHoverOut = this.handleCardHoverOut.bind(this);
  }

  handleCardHover() {
    const {offer, onCardHover} = this.props;
    if (onCardHover) {
      onCardHover(offer);
    }
  }

  handleCardHoverOut() {
    const {onCardHoverOut} = this.props;
    if (onCardHoverOut) {
      onCardHoverOut();
    }
  }

  render() {
    const {offer, type} = this.props;
    const classNamesCard = classnames(`place-card`, {
      'cities__place-card': type === PlaceCardType.CITIES,
      'near-places__card': type === PlaceCardType.NEAR,
      'favorites__card': type === PlaceCardType.FAVORITES,
    });
    const classNamesImgWrap = classnames(`place-card__image-wrapper`, {
      'cities__image-wrapper': type === PlaceCardType.CITIES,
      'near-places__image-wrapper': type === PlaceCardType.NEAR,
      'favorites__image-wrapper': type === PlaceCardType.FAVORITES,
    });
    const classNamesInfo = classnames(`place-card__info`, {
      'favorites__card-info': type === PlaceCardType.FAVORITES,
    });

    return (
      <article className={classNamesCard}
        onMouseOver={this.handleCardHover}
        onMouseOut={this.handleCardHoverOut}
      >
        {offer.isPremium &&
          <div className="place-card__mark">
            <span>Premium</span>
          </div>
        }
        <div className={classNamesImgWrap}>
          <a href="#">
            <img className="place-card__image" src={offer.previewImage} width="260" height="200" alt={offer.title}></img>
          </a>
        </div>
        <div className={classNamesInfo}>
          <div className="place-card__price-wrapper">
            <div className="place-card__price">
              <b className="place-card__price-value">&euro;{offer.price}</b>
              <span className="place-card__price-text">&#47;&nbsp;night</span>
            </div>
            <ButtonFavoriteWrap
              offer={offer}
              type={type}
            />
          </div>
          <div className="place-card__rating rating">
            <div className="place-card__stars rating__stars">
              <span style={{width: `${Math.round(offer.rating) * 20}%`}}></span>
              <span className="visually-hidden">Rating</span>
            </div>
          </div>
          <h2 className="place-card__name">
            <Link to={`${AppRoute.OFFER}/${offer.id}`}>{offer.title}</Link>
          </h2>
          <p className="place-card__type">{HouseTypeTemplate[offer.type]}</p>
        </div>
      </article>
    );
  }
}

export default PlaceCard;
