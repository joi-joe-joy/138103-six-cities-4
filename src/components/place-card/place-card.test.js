import React from "react";
import renderer from "react-test-renderer";
import {PlaceCard} from "./place-card";

describe(`Render PlaceCard`, () => {
  it(`PlaceCard with title`, () => {
    const tree = renderer
      .create(<PlaceCard
        title={`Wood and Rocks`}
        onBookmarkClick={()=>{}}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`PlaceCard without title`, () => {
    const tree = renderer
      .create(<PlaceCard
        title={``}
        onBookmarkClick={()=>{}}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
