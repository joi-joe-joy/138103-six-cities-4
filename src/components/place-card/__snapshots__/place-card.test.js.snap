// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render PlaceCard PlaceCard render correctly with Premium 1`] = `
<article
  className="cities__place-card place-card"
  onMouseOver={[Function]}
>
  <div
    className="place-card__mark"
  >
    <span>
      Premium
    </span>
  </div>
  <div
    className="cities__image-wrapper place-card__image-wrapper"
  >
    <a
      href="#"
    >
      <img
        alt="Beautiful & luxurious apartment at great location"
        className="place-card__image"
        height="200"
        src="img/apartment-01.jpg"
        width="260"
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          380
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button"
        onClick={[Function]}
        type="button"
      >
        <SvgrURL
          height="18"
          width="17"
        />
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        href="#"
      >
        Beautiful & luxurious apartment at great location
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      apartment
    </p>
  </div>
</article>
`;

exports[`Render PlaceCard PlaceCard render correctly without Premium 1`] = `
<article
  className="cities__place-card place-card"
  onMouseOver={[Function]}
>
  <div
    className="place-card__mark"
  >
    <span>
      Premium
    </span>
  </div>
  <div
    className="cities__image-wrapper place-card__image-wrapper"
  >
    <a
      href="#"
    >
      <img
        alt="Beautiful & luxurious apartment at great location"
        className="place-card__image"
        height="200"
        src="img/apartment-01.jpg"
        width="260"
      />
    </a>
  </div>
  <div
    className="place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          380
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button"
        onClick={[Function]}
        type="button"
      >
        <SvgrURL
          height="18"
          width="17"
        />
        <span
          className="visually-hidden"
        >
          To bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        href="#"
      >
        Beautiful & luxurious apartment at great location
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      apartment
    </p>
  </div>
</article>
`;
